library/proc_macro/src/bridge/client.rs

```diff
@@ -366,14 +451,16 @@ fn maybe_install_panic_hook(force_show_panics: bool) {
     });
 }
 
+static SYMBOL_COUNTER: AtomicUsize = AtomicUsize::new(1);
```

Comment:
This may be problematic when using rust-analyzer on 32bit systems as now only 2^32 different idents are possible for the entire time the editor is open. This could be hit when a proc macro randomizes the identifier names. (proc-macros should be deterministic, but are not required to) Rust-analyzer loads the proc macro once at startup and never unloads it.

Reply:
Handles are actually truncated to 32 bits before being used in `OwnedStore`, so this will actually impact both 32-bit and 64-bit hosts similarly:
https://github.com/rust-lang/rust/blob/a8b8558f083d86247ef3260ebb4f97b276cdbf73/library/proc_macro/src/bridge/handle.rs#L29-L30

This was actually already the case before the changes in this patch stack, and this change isn't making anything worse. Previously the code atomic was defined in handle definition macro instead of being written out separately:
https://github.com/rust-lang/rust/blob/a8b8558f083d86247ef3260ebb4f97b276cdbf73/library/proc_macro/src/bridge/client.rs#L23

Whenever a new `Ident` was created it would be intered in this store, and the counter stored in the client dylib's globals would be atomically incremented by `{Interned,Owning}Store`. To my knowledge, @eddyb did this to make sure that if a proc macro stashed an `Ident` (or `Literal`, `Span` or other handle) in global state it would not be possible to observe identifier re-use on subsequent invocations. Because of that, randomized idents won't actually be necessary to exhaust this number, as `*Store` is destroyed after each macro invocation, but the counters are global (within each loaded dylib) so a set of completely fresh integers will be used for the next invocation's store.

Reply:
We could make it thread-local if we're only worried about illegal stashing of handles in proc macro TLS (and not `unsafe` code being used to move them between threads), but then it will still run out unless RA keeps spawning threads.